Assets/__Scripts/MapEditor/Audio/WaveformCache.cs

```diff
@@ -0,0 +1,155 @@
+﻿using System;
+using System.IO;
+using System.IO.Compression;
+using System.Collections;
+using UnityEngine;
+using UnityEngine.Audio;
+using System.Threading;
+using System.Collections.Generic;
+using System.Linq;
+
+public class WaveformCache {
+    public class WaveformData
+    {
+        public int chunks;
+        public int chunksLoaded { get; internal set; }
+        public float[][] bandVolumes { get; internal set; }
+
+        public void initBandVolumes(int len)
+        {
+            if (bandVolumes == null)
+            {
+                bandVolumes = new float[len * chunks][];
+            }
+        }
+
+        public void getChunk(int chunkId, int len, ref float[][] toRender)
+        {
+            Array.Copy(bandVolumes, chunkId * len, toRender, 0, len);
+            Array.Reverse(toRender);
+        }
+    }
+
+    private Thread saveThread;
+    private string fileName;
+    private int ColumnsPerChunk;
+
+    public bool fileHadAllChunks { get; private set; } = false;
+    public WaveformData waveformData { get; private set; } = new WaveformData();
+
+    public WaveformCache(int ColumnsPerChunk) {
+        this.ColumnsPerChunk = ColumnsPerChunk;
+        fileName = Path.Combine(BeatSaberSongContainer.Instance.song.directory, ".waveform");
+    }
+
+    public void ReadWaveformFromFile()
+    {
+        int idx = 0;
+        float[] bandTemp = new float[AudioManager._bands.Length - 1];
+        Debug.Log("WaveformCache: Reading file");
+
+        if (!File.Exists(fileName))
+        {
+            waveformData.chunksLoaded = 0;
+            return;
+        }
+
+        using (FileStream file = File.OpenRead(fileName))
```

Comment:
A thought did cross my mind, and that I do want you to address: What if the audio file is swapped or edited? It'd be a little bit uncommon but I have seen people edit the audio file to adjust the offset, and to also add/remove silent beats from the beginning or end. This would require the waveform cache to essentially be thrown out and be regenerated.

How about maybe adding a way to store and check the hash of the song file? An example of grabbing the hash can be found [here](https://github.com/Caeden117/ChroMapper/blob/dev/Assets/__Scripts/Settings/CustomPlatformSettings.cs#L53).